File: jar-common/src/main/java/com/beancounter/common/model/Market.java
package com.beancounter.common.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import java.util.HashMap;
import java.util.Map;
import java.util.TimeZone;
import javax.persistence.Transient;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.NonNull;
import lombok.ToString;

/**
 * A stock exchange.
 *
 * @author mikeh
 * @since 2019-01-27
 */
@Data
@Builder
@NoArgsConstructor
@AllArgsConstructor
@ToString(onlyExplicitlyIncluded = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Market {
  // ToDo: Separate input props from impl to make it clearer.
  @NonNull
  @ToString.Include
  private String code;
  @Transient
  private Currency currency;
  @JsonIgnore
  private String currencyId;
  @JsonIgnore
  private String timezoneId;
  @Builder.Default
  private TimeZone timezone = TimeZone.getTimeZone("US/Eastern");

  @Builder.Default
  @JsonIgnore
  private Map<String, String> aliases = new HashMap<>();

  @JsonIgnore
  private String enricher;

  @JsonIgnore
  public boolean inMemory() {
    return code.equalsIgnoreCase("MOCK");
  }
}

File: jar-common/src/main/kotlin/com/beancounter/common/model/Market.kt
package com.beancounter.common.model

import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonInclude
import com.fasterxml.jackson.databind.annotation.JsonDeserialize
import com.fasterxml.jackson.databind.annotation.JsonSerialize
import com.fasterxml.jackson.datatype.jsr310.deser.LocalTimeDeserializer
import com.fasterxml.jackson.datatype.jsr310.ser.LocalTimeSerializer
import org.springframework.boot.context.properties.ConstructorBinding
import java.math.BigDecimal
import java.time.LocalTime
import java.util.TimeZone

/**
 * A stock exchange.
 *
 * @author mikeh
 * @since 2019-01-27
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
data class Market @ConstructorBinding constructor(
    val code: String,
    val currencyId: String = "USD",
    val timezoneId: String = "US/Eastern",
    @JsonInclude(JsonInclude.Include.NON_EMPTY)
    val aliases: Map<String, String> = mapOf(),
    val timezone: TimeZone = TimeZone.getTimeZone(timezoneId),
    @Transient
    var currency: Currency = Currency(currencyId),
    @JsonSerialize(using = LocalTimeSerializer::class)
    @JsonDeserialize(using = LocalTimeDeserializer::class)
    @JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "HH:mm")
    val priceTime: LocalTime = LocalTime.of(19, 0),
    val daysToSubtract: Int = 1,
    val enricher: String? = null,
    val multiplier: BigDecimal = BigDecimal("1.0"),
    val type: String = "Public",
)
